jena-fuseki-kafka-module/src/main/java/org/apache/jena/fuseki/kafka/FKBatchProcessor.java

```diff
@@ -0,0 +1,190 @@
+        // TransactionalNull - no batch transaction.
+        if ( transactional == null )
+            transactional = TransactionalNull.create();
+        this.transactional = transactional;
+        this.fkProcessor = fkProcessor;
+    }
+
+    /**
+     * Round the polling loop, updating the record.
+     * Return true if some processing happened.
+     */
+    public boolean receiver(Consumer<String, RequestFK> consumer, DataState dataState, Duration initialPollingDuration) {
+        Objects.requireNonNull(consumer);
+        Objects.requireNonNull(dataState);
+
+        Duration  pollingDuration = initialPollingDuration;
+
+        if ( pollingDuration == null )
+            pollingDuration = FKConst.pollingWaitDuration;
+        final long lastOffsetState = dataState.getOffset();
+        final String topic = dataState.getTopic();
+        try {
+            boolean rtn = false;
+            long commitedState = lastOffsetState;
+            int i;
+            for ( i = 0 ; i < FKConst.MAX_LOOPS_PER_CYCLE ; i++ ) {
+                long newOffset = receiverStep(topic, dataState.getOffset(), consumer, pollingDuration);
+                if ( newOffset == commitedState ) {
+                    // Nothing received.
+                    break;
+                }
+                dataState.setOffset(newOffset);
+                commitedState = newOffset;
+                rtn = true;
+                // Switch to shorter polling wait
+                pollingDuration = FKConst.pollingWaitDurationMore;
+            }
+            if ( LOG.isDebugEnabled() )
+                FmtLog.debug(LOG, "[%s] Exit receiver loop at i=%d", topic, i);
+            return rtn;
+        } catch (RuntimeException ex) {
+            String x = String.format("[%s] %s", dataState.getTopic(), ex.getMessage());
+            Log.error(LOG, x, ex);
+            return false;
+        }
+    }
+
+    private static final boolean VERBOSE = true;
+
+    /** Do one Kafka consumer poll step. */
+    private long receiverStep(String topic, long lastOffsetState, Consumer<String, RequestFK> consumer, Duration pollingDuration) {
+        Objects.requireNonNull(pollingDuration);
+
+        //FmtLog.debug(LOG, "[%s] consumer.poll(%s ms)", topic, pollingDuration.toMillis());
+        ConsumerRecords<String, RequestFK> cRecords = consumer.poll(pollingDuration);
+
+        if ( cRecords.isEmpty() )
+            // Nothing received - no change.
+            return lastOffsetState;
+
+        int count = cRecords.count();
+        int payloadSize = payloadSize(cRecords);
+        Timer timer = batchStart(topic, lastOffsetState, count, payloadSize);
+
+        long newOffset = batchProcess(topic, cRecords);
+
+        // Check expectation.
+        long newOffset2 = lastOffsetState + count;
+        if ( newOffset != newOffset2 )
+            FmtLog.info(LOG, "[%s] Misaligned offsets: [actual=%d, predicated=%d]", topic, newOffset, newOffset2);
```

Comment:
I think there are cases where this might not hold true e.g. reading a topic which has been compacted, reading a topic where the records have been produced transactionally and thus there are control records that won't have been surfaced to us but only used internally by the broker and consumer.

None of these apply currently but perhaps worth a note that this check might be overly strict?

Reply:
Updated.

It is only a log message - the processing uses `newOffset` (last seen offset) regardless.

Reply:
Yup, realised that after I posted the comment